[MLIR][XeGPU] Switch to 1D representation for SIMT code

chencha3 opened the pull request:
As suggested by the name, this PR switch to use the 1D vector to represent SIMT code, thus we can get rid of unnecessary layout attributes in dpas and tensor_desc for SIMT code. This also leads to the relax of verifier for SIMT code, since parameters needed for checking are architecture specific. The correctness check is supposed to be done by a target object.

chencha3 committed:
Switch to 1D representation for SIMT

chencha3 committed:
refine verfier for load_nd and store_nd

charithaintc commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -520,14 +546,41 @@ LogicalResult LoadGatherOp::verify() {
+        return emitOpError() << "doesn't need TransposeAttr for SIMT code";
+    }
+    return success();
+  } else if (valueTy.getRank() == 1 && tdescShape[0] == chunkSize) {
+    // for 1D vector and valueTy.getNumElements() == tdescShape[0] case,
+    // it is a valid SIMT code if chunkSize happens to be the same as
+    // subgroup size, e.g., tensor_desc<16x16xf16, chunkSize = 16>
```
After looking at test cases, it seems like at SIMT level we don't have chunk size info also? any reason for dropping them too?

chencha3 replied:
> why treat this case separately? I think we need to simplify the logic here.
> 
> If we are in SIMT mode `valueTy.rank() == 1` and `valueTy.getNumElems() == chunkSize`
> 
> Or we can separate out the logic for chunkSize == 1 case, which will be more clear.

It seems you are right.

chencha3 replied:
> After looking at test cases, it seems like at SIMT level we don't have chunk size info also? any reason for dropping them too?

We should have it. Could you help to point me the test cases? I need to fix them.

chencha3 replied:
> why treat this case separately? I think we need to simplify the logic here.
> 
> If we are in SIMT mode `valueTy.rank() == 1` and `valueTy.getNumElems() == chunkSize`
> 
> Or we can separate out the logic for chunkSize == 1 case, which will be more clear.

I think I have fixed this, could you help to have a check?

charithaintc replied:
it looks good now. Thanks!

Garra1980 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -302,9 +270,33 @@ LogicalResult LoadNdOp::verify() {
+  }
+
+  // Check SIMD mode.
   // adjusted tensor descriptor shape tracks the expected shape of the result.
```
since name has been changed let's remove 'adjusted' from the comment

chencha3 replied:
fixed

chencha3 committed:
refine verifier for gather/scatter

chencha3 committed:
update comments

chencha3 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:
```diff
@@ -336,32 +337,30 @@ LogicalResult TensorDescType::verify(
     auto chunkSize = scatterAttr.getChunkSize().getInt();
     // Verify if the first dimension of the tensor descriptor shape is
     // distributable.
-    assert(tdescShape[0] % (laneLayout[0]) == 0 &&
```
I think it is a small issue after confirming with Charitha. tdescShape[0] has to be equal to laneLayout[0], such that a SIMD instruction is dispatched into a SIMT instruction. if tdescShape[0] % lane_layout[0] == 0, it will imply a SIMD instruction could be dispatched into multiple SIMT instructions, which is actually part of logic of blocking.

chencha3 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -559,14 +612,42 @@ LogicalResult StoreScatterOp::verify() {
+    // for 1D vector and valueTy.getNumElements() == tdescShape[0] case,
+    // it is a valid SIMT code if chunkSize happens to be the same as
+    // subgroup size, e.g., tensor_desc<16x16xf16, chunkSize = 16>
+    return success();
```
fixed.

chencha3 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:
```diff
@@ -336,32 +337,30 @@ LogicalResult TensorDescType::verify(
 //        [n_distribution_units, lane_data_size]
```
Updated

charithaintc commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -559,14 +600,31 @@ LogicalResult StoreScatterOp::verify() {
+  // a valid shape for SIMT case
+  if (valueTy.getRank() == 1 && valueTy.getNumElements() == chunkSize) {
+    if (tdescTy.getLayoutAttr())
+      return emitOpError()
+             << "TensorDesc doesn't need LayoutAttr for SIMT code";
+    if (getTransposeAttr())
+      return emitOpError() << "doesn't need TransposeAttr for SIMT code";
+    return success();
+  }
+
+  if (tdescTy.getRank() == 2 && valueTy.getRank() == 2) {
     if (!getTransposeAttr())
       return emitOpError("Store of a rank-2 tensor has to be transposed.");
     transpose({1, 0}, tdescShape);
   }
 
-  return isArgShapesValid(tdescTy, valueTy, tdescShape,
-                          [&]() { return emitOpError(); });
+  if (tdescShape != valueShape)
+    return emitOpError() << "Value shape " << makeString(valueShape)
+                         << " is neither a valid distribution for SIMT nor "
+                            "consistent with the tensor descriptor for SIMD "
+                         << tdescTy;
+
+  return success();
```
seems like this part of the code is same for load_gather and store_scatter? maybe refactoring would be nice.

chencha3 replied:
Fixed

adam-smnk commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -602,51 +660,16 @@ LogicalResult DpasOp::verify() {
-    }
-    return !result;
-  };
+  if (getAcc()) {
+    if (getAcc().getType() != getResultType())
```
nit: could be merged into single if

chencha3 replied:
fixed

adam-smnk commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -602,51 +660,16 @@ LogicalResult DpasOp::verify() {
+  // SIMT code: skip the check since lack of semantic info at this level.
+  // Users need to ensure the correctness.
+  if (lhsRank == 1 && rhsRank == 1 && resRank == 1) {
+    return success();
```
Could we still apply some sanity checks? Maybe that the sizes are at least power of 2 as I expect (but not sure) that's a thing.

chencha3 replied:
> Could we still apply some sanity checks? Maybe that the sizes are at least power of 2 as I expect (but not sure) that's a thing.

It seems there not too much common features across architectures. The one I found is B operand is 32-bit aligned.

adam-smnk reviewed (commented):
Looks pretty good 👍
I really like how much it simplified both in code and IR - now SIMD vs SIMT is more obvious.

Left only minor nits

chencha3 committed:
refator verifiers for load_gather, store_scatter and dpas

chencha3 committed:
fix format

chencha3 committed:
Merge branch 'main' into xegpu_simt_format

chencha3 commented:
> Looks pretty good 👍 I really like how much it simplified both in code and IR - now SIMD vs SIMT is more obvious.
> 
> Left only minor nits

Thanks. I have partially addressed your comments.

chencha3 committed:
address comments

chencha3 commented on mlir/lib/Dialect/XeGPU/IR/XeGPUOps.cpp:
```diff
@@ -602,51 +633,20 @@ LogicalResult DpasOp::verify() {
+    if (numElems % factor != 0)
+      return emitOpError("Expecting B operand to be a multiple of 32 bits.");
+    return success();
+  } else { // SIMD code
```
Yes, fixed
